mdp/nodes/sfa_nodes.py

```diff
@@ -181,39 +258,108 @@ def _execute(self, x, n=None):
             bias = self._bias
         return mult(x, sf) - bias
 
-    def _inverse(self, y):
-        return mult(y, pinv(self.sf)) + self.avg
+    def _svd_solver(self, rng):
+        """
+        Alternative routine to solve the SFA eigenvalue issue. This can be used
+        in case the normal symeig() call in _stop_training() throws the common
+        SymeigException ('Covariance matrices may be singular').
 
-    def get_eta_values(self, t=1):
-        """Return the eta values of the slow components learned during
-        the training phase. If the training phase has not been completed
-        yet, call `stop_training`.
+        For details on the used algorithm see:
+            http://www.geo.tuwien.ac.at/downloads/tm/svd.pdf (section 0.3.2)
+        """
 
-        The delta value of a signal is a measure of its temporal
-        variation, and is defined as the mean of the derivative squared,
-        i.e. delta(x) = mean(dx/dt(t)^2).  delta(x) is zero if
-        x is a constant signal, and increases if the temporal variation
-        of the signal is bigger.
+        U, s, _ = np.linalg.svd(self.cov_mtx)
+        X1 = np.dot(U, np.diag(1.0 / s ** 0.5))
+        X2, _, _ = np.linalg.svd(np.dot(X1.T, np.dot(self.dcov_mtx, X1)))
```

Comment:
Have you checked with Laurenz that this is meaningful? SFA looks for the smallest eigenvalues of this matrix, and SVD clips eigenvalues that are below a certain threshold, so you might in principle discarding the slowest components.

Reply:
It appears to me like numx' svd actually returns U, s, V of full size, but containing some areas of zeros to represent the rank deficit. So I suppose there should be done some slicing here to s and U, see [this draft](https://github.com/mdp-toolkit/mdp-toolkit/pull/31/files#diff-e0a5be9e597a4c5d554db9950e4155edR364).

Reply:
@pberkes No. In fact I've been asking around for someone more comfortable with the math to take a look at this SVD-based method to make sure this does what we want it to do. I have, however, compared the results of the SVD solution to the _symeig solution (when the latter runs successfully) and the final plots of my experiments are essentially identical.